src/audio/smart_amp_test.c

```diff
@@ -0,0 +1,722 @@
+	sad->model.data_size = size;
+	sad->model.data_pos = 0;
+
+	return 0;
+}
+
+static struct comp_dev *smart_amp_new(const struct comp_driver *drv,
+				      struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *sa;
+	struct sof_ipc_comp_process *ipc_sa =
+		(struct sof_ipc_comp_process *)comp;
+	struct smart_amp_data *sad;
+	struct sof_smart_amp_config *cfg;
+	size_t bs;
+	int ret;
+
+	comp_cl_info(&comp_smart_amp, "smart_amp_new()");
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+	dev->drv = drv;
+
+	sa = COMP_GET_IPC(dev, sof_ipc_comp_process);
+
+	ret = memcpy_s(sa, sizeof(*sa), ipc_sa,
+		       sizeof(struct sof_ipc_comp_process));
+	assert(!ret);
+
+	sad = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*sad));
+
+	if (!sad) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, sad);
+
+	cfg = (struct sof_smart_amp_config *)ipc_sa->data;
+	bs = ipc_sa->size;
+
+	if ((bs > 0) && (bs < sizeof(struct sof_smart_amp_config))) {
+		comp_err(dev, "smart_amp_new(): failed to apply config");
+
+		if (sad)
+			rfree(sad);
+		rfree(sad);
+		return NULL;
+	}
+
+	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg, bs);
+
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+}
+
+static int smart_amp_set_config(struct comp_dev *dev,
+				struct sof_ipc_ctrl_data *cdata)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct sof_smart_amp_config *cfg;
+	size_t bs;
+
+	comp_info(dev, "smart_amp_set_config()");
+
+	/* Copy new config, find size from header */
+	cfg = (struct sof_smart_amp_config *)cdata->data->data;
+	bs = cfg->size;
+
+	comp_info(dev, "smart_amp_set_config(), blob size = %u", bs);
+
+	if (bs != sizeof(struct sof_smart_amp_config)) {
+		comp_err(dev, "smart_amp_set_config(): invalid blob size");
+		return -EINVAL;
+	}
+
+	memcpy_s(&sad->config, sizeof(struct sof_smart_amp_config), cfg,
+		 sizeof(struct sof_smart_amp_config));
+
+	return 0;
+}
+
+static int smart_amp_get_config(struct comp_dev *dev,
+				struct sof_ipc_ctrl_data *cdata, int size)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	size_t bs;
+	int ret = 0;
+
+	comp_info(dev, "smart_amp_get_config()");
+
+	// Copy back to user space
+	bs = sad->config.size;
+	comp_info(dev, "smart_amp_get_config(): block size: %u", bs);
+
+	if (bs == 0 || bs > size)
+		return -EINVAL;
+
+	ret = memcpy_s(cdata->data->data, size, &sad->config, bs);
+	assert(!ret);
+
+	cdata->data->abi = SOF_ABI_VERSION;
+	cdata->data->size = bs;
+
+	return ret;
+}
+
+static int smart_amp_get_model(struct comp_dev *dev,
+			       struct sof_ipc_ctrl_data *cdata, int size)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	size_t bs;
+	int ret = 0;
+
+	comp_dbg(dev, "smart_amp_get_model() msg_index = %d, num_elems = %d, remaining = %d ",
+		 cdata->msg_index, cdata->num_elems,
+		 cdata->elems_remaining);
+
+	/* Copy back to user space */
+	if (sad->model.data) {
```

Comment:
This split/concatenate code for large parameters seems to be duplicated here and in a couple of other components, looks like a very good candidate for a common set of helper functions in order to enable quick reuse by others.

Reply:
I've rewrite those functions and placed them in component.c as generic (also refined detect_test component to use them).